error.go

```diff
@@ -51,19 +54,64 @@ func firstError(a, b error) error {
 	return b
 }
 
+const (
+	hint          = 0x0001
+	positionStart = 0xfff1
+	lineStart     = 0xfff3
+)
+
+func atoiOrPanic(s string) int {
+	i, err := strconv.Atoi(s)
+	if err != nil {
+		panic(err)
+	}
+
+	return i
+}
+
 // decodeError decodes an error response
 // https://www.edgedb.com/docs/internals/protocol/messages#errorresponse
-func decodeError(r *buff.Reader) error {
+func decodeError(r *buff.Reader, query string) error {
 	r.Discard(1) // severity
-	err := errorFromCode(r.PopUint32(), r.PopString())
-	n := int(r.PopUint16())
+	code := r.PopUint32()
+	msg := r.PopString()
 
+	n := int(r.PopUint16())
+	headers := make(map[uint16]string, n)
 	for i := 0; i < n; i++ {
-		r.PopUint16() // key
-		r.PopString() // value
+		headers[r.PopUint16()] = r.PopString()
 	}
 
-	return err
+	// if the error is syntax related
+	if 0x04_01_00_00 <= code && code < 0x04_02_00_00 {
```

Comment:
Well, all kinds of errors can have position information in them. IMO we should instead check for that, and then render a nice message for all kinds of errors. E.g. a query might be correct syntactically, but have an invalid type reference; the server would return the exact spot of the problem in the query.

Reply:
changed